File: server/src/main/java/com/finance/tracker/service/ExpenseService.java
package com.finance.tracker.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Service;

import com.finance.tracker.exception.ResourceNotFoundException;
import com.finance.tracker.model.Expense;
import com.finance.tracker.repository.ExpenseRepository;

@Service
public class ExpenseService {

    @Autowired
    private ExpenseRepository repository;

    public Expense addExpense(@NonNull Expense expense) {
        return repository.save(expense);
    }

    public List<Expense> getExpensesByCategory(String category) {
        return repository.findByCategory(category);
    }

    public List<Expense> getExpensesByDate(String date) {
        return repository.findByDate(date);
    }

    public Expense updateExpense(@NonNull String id, Expense expense) {
        Expense existing = repository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Expense not found with id: " + id));

        existing.setTitle(expense.getTitle());
        existing.setAmount(expense.getAmount());
        existing.setCategory(expense.getCategory());
        existing.setDate(expense.getDate());

        return repository.save(existing);
    }

    public void deleteExpense(@NonNull String id) {
        if (!repository.existsById(id)) {
            throw new ResourceNotFoundException("Expense not found with id: " + id);
        }
        repository.deleteById(id);
    }

    public Expense getExpenseById(@NonNull String id) {
        return repository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Expense not found with id: " + id));
    }

    public List<Expense> getAllExpenses() {
        return repository.findAll();
    }
}

File: client/src/services/expenseService.js
import API from "./api";

export const getExpenses = () => API.get("/expenses");

export const addExpense = (data) => API.post("/expenses", data);

export const deleteExpense = (id) => API.delete(`/expenses/${id}`);

export const updateExpense = (id, data) => API.put(`/expenses/${id}`, data);

export const getExpensesByDate = (date) => API.get(`/expenses/date/${date}`);

export const getExpensesByCategory = (category) => API.get(`/expenses/category/${category}`);

export const getFilteredExpenses = (date, category) => API.get(`/expenses`, {params: { date, category }});